ccp.c

```diff
@@ -246,42 +250,48 @@ int ccp_read_msg(
         PRINT("unknown connection: %u\n", hdr.SocketId);
         return -5;
     }
+    msg_ptr = buf + ok;
 
     state = get_ccp_priv_state(conn);
     if (hdr.Type == INSTALL_EXPR) {
-        memset(&emsg, 0, sizeof(struct InstallExpressionMsg));
-        ok = read_install_expr_msg(&hdr, &emsg, buf + ok);
+        memset(&expr_msg_info, 0, sizeof(struct InstallExpressionMsgHdr));
+        ok = read_install_expr_msg_hdr(&hdr, &expr_msg_info, msg_ptr); 
         if (ok < 0) {
-            PRINT("could not read install expression msg: %d\n", ok);
+            PRINT("could not read install expression msg header: %d\n", ok);
             return -6;
         }
+        msg_ptr += ok;
 
         ACQUIRE_LOCK(&ccp_state_lock);
         memset(state->expressions, 0, MAX_EXPRESSIONS * sizeof(struct Expression));
         memset(state->fold_instructions, 0, MAX_INSTRUCTIONS * sizeof(struct Instruction64));
     
-        state->program_uid = emsg.program_uid;
-        state->num_expressions = emsg.num_expressions;
-        state->num_instructions = emsg.num_instructions;
+        state->program_uid = expr_msg_info.program_uid;
+        state->num_expressions = expr_msg_info.num_expressions;
+        state->num_instructions = expr_msg_info.num_instructions;
 
         // parse the expressions 
         for (i=0; i<state->num_expressions; i++) {
-            ok = read_expression(&(state->expressions[i]), &(emsg.exprs[i]));
+            current_expr = (struct ExpressionMsg*)(msg_ptr);
+            ok = read_expression(&(state->expressions[i]), current_expr);
             if (ok < 0) {
-                PRINT("could not read expression\n");
+                PRINT("could not read expression # %lu, error: %d\n", i, ok);
                 RELEASE_LOCK(&ccp_state_lock);
                 return -7;
             }
+            msg_ptr += sizeof(struct ExpressionMsg);
         }
 
         // parse the instructions
         for (i=0; i<state->num_instructions; i++) {
```

Comment:
same thing here replacing the loop with a memcpy

Reply:
You can't do a memcpy here. If you look at read_instruction(), you actually need to do parsing work (it's not just copying the memory over).

Reply:
It works in the expression case because the way they are sent down is exactly how they're laid out in memory in libccp